File: 백준/Gold/5427. 불/불.py
from collections import deque
def bfs():
    time  = 0
    global ans

    while q:
        time += 1
        # 불 1초동안 확장
        for _ in range(len(fires)):
            x, y = fires.popleft()
            for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1)):
                nx, ny = x + dx, y + dy
                if 0 <= nx < n and 0 <= ny < m and (arr[nx][ny] == "." or arr[nx][ny] =="@") :
                    arr[nx][ny] = "*"
                    fires.append((nx, ny))

        # 사람 이동
        for _ in range(len(q)):
            x, y = q.popleft()
            for dx, dy in ((-1, 0), (1, 0), (0, -1), (0, 1)):
                nx, ny = x + dx, y + dy
                if 0 <= nx < n and 0 <= ny < m:
                    if arr[nx][ny] == ".":
                        arr[nx][ny] = "@"
                        q.append((nx, ny))
                # 범위 탈출조건
                else:
                    ans = time
                    return



for t in range(int(input())):
    m,n = map(int,input().split())
    arr = [list(input()) for _ in range(n)]
    v = [[-1] * m for _ in range(n)]
    visit = [[0] * m for _ in range(n)]
    ans ="IMPOSSIBLE"
    fires = deque()
    q = deque()

    for i in range(n):
        for j in range(m):
            if arr[i][j] == '*':
                fires.append((i,j))
            elif arr[i][j] == '@':
                q.append((i,j))
    bfs()
    print(ans)

File: 백준/Gold/5427. 불/불.java
import java.util.*;
import java.io.*;

public class Main {
    static int n, m;
    static char[][] arr;
    static Deque<int[]> fires;
    static Deque<int[]> q;
    static String ans;

    static void bfs() {
        int time = 0;

        while (!q.isEmpty()) {
            time++;
            
            int fireSize = fires.size();
            for (int i = 0; i < fireSize; i++) {
                int[] cur = fires.poll();
                int x = cur[0], y = cur[1];

                int[] dx = {-1, 1, 0, 0};
                int[] dy = {0, 0, -1, 1};

                for (int k = 0; k < 4; k++) {
                    int nx = x + dx[k];
                    int ny = y + dy[k];

                    if (0 <= nx && nx < n && 0 <= ny && ny < m &&
                            (arr[nx][ny] == '.' || arr[nx][ny] == '@')) {
                        arr[nx][ny] = '*';
                        fires.add(new int[]{nx, ny});
                    }
                }
            }

            int playerSize = q.size();
            for (int i = 0; i < playerSize; i++) {
                int[] cur = q.poll();
                int x = cur[0], y = cur[1];

                int[] dx = {-1, 1, 0, 0};
                int[] dy = {0, 0, -1, 1};

                for (int k = 0; k < 4; k++) {
                    int nx = x + dx[k];
                    int ny = y + dy[k];

                    // 맵 밖 → 탈출 성공
                    if (nx < 0 || nx >= n || ny < 0 || ny >= m) {
                        ans = String.valueOf(time);
                        return;
                    }

                    // 이동 가능
                    if (arr[nx][ny] == '.') {
                        arr[nx][ny] = '@';
                        q.add(new int[]{nx, ny});
                    }
                }
            }
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int T = Integer.parseInt(br.readLine());

        while (T-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            m = Integer.parseInt(st.nextToken());  // width
            n = Integer.parseInt(st.nextToken());  // height

            arr = new char[n][m];
            fires = new LinkedList<>();
            q = new LinkedList<>();

            for (int i = 0; i < n; i++) {
                String line = br.readLine();
                for (int j = 0; j < m; j++) {
                    arr[i][j] = line.charAt(j);
                    if (arr[i][j] == '*') fires.add(new int[]{i, j});
                    else if (arr[i][j] == '@') q.add(new int[]{i, j});
                }
            }

            ans = "IMPOSSIBLE";
            bfs();
            sb.append(ans).append("\n");
        }

        System.out.print(sb);
    }
}
